resources/sqlite/src/lib.rs

```diff
@@ -0,0 +1,154 @@
+//! #[shuttle_runtime::main]
+//! async fn axum(
+//!     #[shuttle_sqlite::ShuttleSqlite] pool: shuttle_sqlite::SqlitePool,
+//! ) -> shuttle_axum::ShuttleAxum {
+//!     let _ = sqlx::query(
+//!         "CREATE TABLE IF NOT EXISTS users(id int, name varchar(128), email varchar(128));",
+//!     )
+//!     .execute(&pool)
+//!     .await
+//!     .unwrap();
+//! }
+//! ```
+//!
+//! ## Configuration
+//! The database can be configured using [`ShuttleSqliteConnOpts`] which mirrors sqlx's [`SqliteConnectOptions`](https://docs.rs/sqlx/latest/sqlx/sqlite/struct.SqliteConnectOptions.html) for the
+//! options it exposes.
+//!
+//! Construction of the full connection string is handled internally for security reasons and defaults to creating a
+//! file-based database named `default_db.sqlite` with `create_if_missing == true`. Use the `filename` and/or
+//! `in_memory` methods to configure the type of database created.
+//!
+//! See [`SqliteConnectOptions::new()`](https://docs.rs/sqlx/latest/sqlx/sqlite/struct.SqliteConnectOptions.html#method.new)
+//! for all default settings.
+//!
+//! ```ignore
+//! #[shuttle_runtime::main]
+//! async fn axum(
+//!     #[shuttle_sqlite::ShuttleSqlite(opts = ShuttleSqliteConnOpts::new().filename("custom.sqlite"))] pool: shuttle_sqlite::SqlitePool,
+//! ) -> shuttle_axum::ShuttleAxum { /* ... */ }
+//! ```
+//! Note that Shuttle does currently not support the `collation`, `thread_name`, `log_settings`, `pragma`, `extension`,
+//! `shared_cache` options.
+use std::path::Path;
+
+use async_trait::async_trait;
+use serde::Serialize;
+use shuttle_service::{Factory, ResourceBuilder, Type};
+
+mod conn_opts;
+pub use conn_opts::*;
+
+use sqlx::sqlite::SqliteConnectOptions;
+
+/// Builder struct used to configure the database, e.g. `SQLite(opts = SQLiteConnOpts::new())`.
+#[derive(Serialize)]
+pub struct ShuttleSqlite {
+    opts: ShuttleSqliteConnOpts,
+}
+
+impl ShuttleSqlite {
+    pub fn opts(mut self, opts: ShuttleSqliteConnOpts) -> Self {
+        self.opts = opts;
+        self
+    }
+
+    pub fn filename(mut self, filename: impl AsRef<Path>) -> Self {
+        self.opts = self.opts.filename(filename);
+        self
+    }
+
+    pub fn in_memory(mut self, on: bool) -> Self {
+        self.opts.in_memory = on;
+        self
+    }
+}
+
+#[async_trait]
+impl ResourceBuilder<sqlx::SqlitePool> for ShuttleSqlite {
+    const TYPE: Type = Type::EmbeddedDatabase;
+
+    type Config = Self;
+
+    type Output = ShuttleSqliteConnOpts;
+
+    fn new() -> Self {
+        Self {
+            opts: ShuttleSqliteConnOpts::default(),
+        }
+    }
+
+    fn config(&self) -> &Self::Config {
+        &self
+    }
+
+    async fn output(
+        mut self,
+        factory: &mut dyn Factory,
+    ) -> Result<Self::Output, shuttle_service::Error> {
+        // We construct an absolute path using `storage_path` to prevent user access to other parts of the file system.
+        let storage_path = factory.get_storage_path()?;
```

Comment:
I have some major refactorings of the deployer storage system in sight with #1050. Although that is far from complete, it would probably be simpler if this was delayed until after #1050 (and the Beta arch in general) to not complicate that rewrite even more. WDYT @chesedo @oddgrd ?

Reply:
If I may jump in here, I would love to have this merged beforehand as that plan sounds like it would make some heavy refactorings necessary for which I don't really have much time right now. This has been gotten out of sync a few times and required some updates which already took a lot longer than I would have preferred.

Reply:
I know it sucks to have a huge PR waiting around 😞. I'll see if we can escalate the priority of this one.